packages/rum-core/src/domain/contexts/pageStateContexts.spec.ts

```diff
@@ -0,0 +1,113 @@
+import type { Duration, RelativeTime } from '@datadog/browser-core'
+import { resetExperimentalFeatures, updateExperimentalFeatures } from '@datadog/browser-core'
+import type { TestSetupBuilder } from 'packages/rum-core/test/specHelper'
+import { setup } from '../../../test/specHelper'
+import type { PageStateContexts } from './pageStateContexts'
+import { resetPageStates, startPageStateContexts, addPageState, PageState } from './pageStateContexts'
+
+describe('pageContexts', () => {
+  let pageStateContexts: PageStateContexts
+  let setupBuilder: TestSetupBuilder
+
+  beforeEach(() => {
+    updateExperimentalFeatures(['resource_page_states'])
+    setupBuilder = setup()
+      .withFakeClock()
+      .beforeBuild(() => {
+        pageStateContexts = startPageStateContexts()
+        return pageStateContexts
+      })
+  })
+
+  afterEach(() => {
+    setupBuilder.cleanup()
+    resetPageStates()
+    resetExperimentalFeatures()
+  })
+
+  it('should not track page states when ff resource_page_states is disabled', () => {
+    resetExperimentalFeatures()
+    setupBuilder.build()
+    expect(pageStateContexts.getPageStates(0 as RelativeTime, 10 as RelativeTime)).not.toBeDefined()
+  })
+
+  it('should have the current state when starting', () => {
+    setupBuilder.build()
+    expect(pageStateContexts.getPageStates(0 as RelativeTime, 10 as RelativeTime)).toBeDefined()
+  })
+
+  it('should return undefined if timings out of timeline bounds', () => {
+    pageStateContexts = startPageStateContexts()
+    expect(pageStateContexts.getPageStates(-10 as RelativeTime, 0 as RelativeTime)).not.toBeDefined()
+  })
+
+  it('should track the number and duration of all page state', () => {
+    const { clock } = setupBuilder.build()
+    resetPageStates()
+
+    clock.tick(0)
+    addPageState(PageState.ACTIVE)
+
+    clock.tick(10)
+    addPageState(PageState.PASSIVE)
+
+    clock.tick(10)
+    addPageState(PageState.HIDDEN)
+
+    clock.tick(10)
+    addPageState(PageState.FROZEN)
+
+    expect(pageStateContexts.getPageStates(0 as RelativeTime, 50 as RelativeTime)).toEqual({
+      active: { count: 1, duration: 10 as Duration },
+      passive: { count: 1, duration: 10 as Duration },
+      hidden: { count: 1, duration: 10 as Duration },
+      frozen: { count: 1, duration: 20 as Duration },
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Not sure why `frozen` duration is 20 in this case?

Reply:
In the timeline I got:
1. from 0 to 10 => state = active
2. from 10 to 20 => state = passive
3. from 20 to 30 => state = hidden
4. from 30 to infinity => state = frozen

With  `getPageStates(0 as RelativeTime, 50 as RelativeTime))`  from 30 to 50 I get `frozen`

Reply:
Ooh right, thank you :)